File: continuum-js/continuum-idl-js/src/api/HasDecorators.ts
import {C3Decorator} from '@/api/decorators/C3Decorator'

/**
 * Interface for classes that have {@link C3Decorator}s.
 */
export interface HasDecorators {

    decorators?: C3Decorator[] | null

    /**
     * Checks if this type has a {@link C3Decorator} for the given type
     * @param value the {@link C3Decorator} to check for
     * @return true if the type has a {@link C3Decorator} for the given type false if not
     */
    containsDecorator(value: C3Decorator): boolean

    /**
     * Checks if this type has any {@link C3Decorator}
     * @return true if any {@link C3Decorator}s are present false if not
     */
    hasDecorators(): boolean

    /**
     * Adds a new decorator to this type
     * @param decorator the decorator to add
     * @return this {@link C3Base}
     */
    addDecorator(decorator: C3Decorator): void

    /**
     * Finds a decorator for the given type if it exists
     * @param value the type to find
     * @return the decorator if it exists null if not
     */
    findDecorator(value: C3Decorator): C3Decorator | null

}

File: continuum-idl/src/main/java/org/kinotic/continuum/idl/api/schema/HasDecorators.java
package org.kinotic.continuum.idl.api.schema;

import org.kinotic.continuum.idl.api.schema.decorators.C3Decorator;

import java.util.List;

/**
 * Defines a type that can have {@link C3Decorator}s
 * Created by Navíd Mitchell 🤪 on 4/19/24.
 */
public interface HasDecorators {

    /**
     * Checks if this type contains a {@link C3Decorator} of the given subclass
     *
     * @param clazz to see if this type has
     * @return true if the {@link C3Decorator} is present false if not
     */
    default boolean containsDecorator(Class<? extends C3Decorator> clazz){
        return findDecorator(clazz) != null;
    }

    /**
     * Checks if this type has any {@link C3Decorator}
     *
     * @return true if any {@link C3Decorator}s are present false if not
     */
    default boolean hasDecorators(){
        return getDecorators() != null && !getDecorators().isEmpty();
    }

    /**
     * Finds the first {@link C3Decorator} of the given subclass or null if none are found
     *
     * @param clazz to find the {@link C3Decorator} for
     * @return the {@link C3Decorator} or null if none are found
     */
    default <T extends C3Decorator> T findDecorator(Class<T> clazz){
        T ret = null;
        if(getDecorators() != null){
            for (C3Decorator decorator : getDecorators()){
                if(clazz.isAssignableFrom(decorator.getClass())){
                    //noinspection unchecked
                    ret = (T) decorator;
                    break;
                }
            }
        }
        return ret;
    }

    /**
     * Gets all {@link C3Decorator}s for this type
     *
     * @return all {@link C3Decorator}s for this type
     */
    List<C3Decorator> getDecorators();

}
